diff --git a/src/init.luau b/src/init.luau
index 11d1bd2..d1e68d1 100644
--- a/src/init.luau
+++ b/src/init.luau
@@ -2,9 +2,8 @@ assert(game, "when using vide outside of Roblox, require lib.luau instead")
 
 local vide = require(script.lib)
 
-export type source<T> = vide.source<T>
 export type Source<T> = vide.Source<T>
-export type context<T> = vide.context<T>
 export type Context<T> = vide.Context<T>
+export type Derivable<T> = T | () -> T
 
 return vide
