diff --git a/src/anim/spring/spring_file.luau b/src/anim/spring/spring_file.luau
index 68a0363..e7ea1e3 100644
--- a/src/anim/spring/spring_file.luau
+++ b/src/anim/spring/spring_file.luau
@@ -58,6 +58,10 @@ local function pack_type(value: AnimatableDataType): (AnimatableDataTypeEnum, { 
 		return "vector", { value.x, value.y, value.z }
 	elseif type(value) == "number" then
 		return "number", { value }
+	elseif typeof(value) == "Vector2" then
+		return "Vector2", { value.X, value.Y }
+	elseif typeof(value) == "Vector3" then
+		return "Vector3", { value.X, value.Y, value.Z }
 	end
 
 	error(`tried tweening {typeof(value)}`)
@@ -84,6 +88,10 @@ local function unpack_type(
 		return UDim2.new(array[1], array[2], array[3], array[4])
 	elseif type_string == "number" then
 		return array[1]
+	elseif type_string == "Vector2" then
+		return Vector2.new(array[1], array[2])
+	elseif type_string == "Vector3" then
+		return Vector3.new(array[1], array[2], array[3])
 	end
 
 	return error("unreachable")
